WalletWasabi.Tests/UnitTests/BlockchainAnalysis/SelfSpendAnonScoreTests.cs

```diff
@@ -0,0 +1,95 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using WalletWasabi.Tests.Helpers;
+using Xunit;
+
+namespace WalletWasabi.Tests.UnitTests.BlockchainAnalysis
+{
+	/// <summary>
+	/// In these tests all the inputs and outputs of a transaction are controlled by the user.
+	/// </summary>
+	public class SelfSpendAnonScoreTests
+	{
+		[Fact]
+		public void OneOwnInOneOwnOut()
+		{
+			var analyser = ServiceFactory.BlockchainAnalyzer();
+			var tx = BitcoinFactory.SmartTransaction(0, 0, 1, 1);
+			var coin = Assert.Single(tx.WalletInputs);
+			var key = coin.HdPubKey;
+			key.AnonymitySet = 3;
+
+			analyser.Analyze(tx);
+
+			// Anonset of the input shall be retained.
+			// Although the tx has more than one interpretation
+			// blockchain anal usually just assumes it's a self spend.
+			Assert.All(tx.WalletOutputs, x => Assert.Equal(3, x.HdPubKey.AnonymitySet));
+			Assert.All(tx.WalletInputs, x => Assert.Equal(3, x.HdPubKey.AnonymitySet));
+		}
+
+		[Fact]
+		public void OneOwnInManyOwnOut()
+		{
+			var analyser = ServiceFactory.BlockchainAnalyzer();
+			var tx = BitcoinFactory.SmartTransaction(0, 0, 1, 3);
+			var coin = Assert.Single(tx.WalletInputs);
+			var key = coin.HdPubKey;
+			key.AnonymitySet = 3;
+
+			analyser.Analyze(tx);
+
+			// Anonset of the input shall be retained.
+			// Although the tx has many interpretations we shall not guess which one
+			// a blockchain analyzer would go with, therefore outputs shall not gain anonsets
+			// as we're conservatively estimating.
+			Assert.All(tx.WalletOutputs, x => Assert.Equal(3, x.HdPubKey.AnonymitySet));
+			Assert.All(tx.WalletInputs, x => Assert.Equal(3, x.HdPubKey.AnonymitySet));
+		}
+
+		[Fact]
+		public void ManyOwnInOneOwnOut()
+		{
+			var analyser = ServiceFactory.BlockchainAnalyzer();
+			var tx = BitcoinFactory.SmartTransaction(0, 0, 3, 1);
+			var smallestAnonset = 3;
+
+			tx.WalletInputs.First().HdPubKey.AnonymitySet = smallestAnonset;
+			foreach (var coin in tx.WalletInputs.Skip(1))
+			{
+				coin.HdPubKey.AnonymitySet = 100;
+			}
+
+			analyser.Analyze(tx);
+
+			// Anonset of the input shall be worsened because of input merging.
+			Assert.All(tx.WalletOutputs, x => Assert.True(x.HdPubKey.AnonymitySet < smallestAnonset));
```

Comment:
Why not take the smallest one?

Reply:
I am either misunderstanding you or you forgot something since you're the one who implemented consolidation penalty for the first time: https://github.com/zkSNACKs/WalletWasabi/pull/4631

Reply:
Discussed it IRL. It is OK.